### PR TITLE
Lwsf ndp

### PR DESCRIPTION
**Pull request by kbara**
This commit adds two types of NDP support:
- One listens for neighbor solicitations and replies with advertisements.
  It is always on.
- The other is implicitly on if the mac address of the next hop on the IPv6
  side is not set. It sends out a request for it on start.

Documentation and end-to-end tests are included.

### CODE REVIEW

File: src/apps/lwaftr/ipv6_apps.lua [hunk 1]
```diff
@@ -147,3 +150,65 @@ function Fragmenter:push ()
...
+      -- configuration reloads?
+      -- This would be an optimization, not a correctness issue
+   end
+   for _=1,link.nreadable(isouth) do
+      local p = receive(isouth)
+      if ndp.is_ndp(p) then
+         if not self.dst_eth and ndp.is_solicited_neighbor_advertisement(p) then
+            local dst_ethernet = ndp.get_dst_ethernet(p, {self.conf.dst_ipv6})
+            if dst_ethernet then
+               self.conf.dst_eth = dst_ethernet
```
Comment (wingo) [inline src/apps/lwaftr/ipv6_apps.lua]: Can we avoid updating the config, and instead update self.dst_eth or so?

Other than this comment, this module looks really nice :-))

Comment (kbara) [inline src/apps/lwaftr/ipv6_apps.lua]: Sure, will do. Thanks. :)

File: src/apps/lwaftr/ndp.lua [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+end
+
+local function icmpv6_type_is_ns(pkt)
+   return pkt.data[eth_ipv6_size] == icmpv6_ns
+end
+
+-- The relevant byte is:
+-- [ R S O 0 0 0 0 0 0 ], where S = solicited
+local function is_solicited_na(pkt)
+   local offset = eth_ipv6_size + 4
```
Comment (wingo) [inline src/apps/lwaftr/ndp.lua]: can we name these constants please.  tx!

File: src/apps/lwaftr/ndp.lua [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+   return bit.band(0x40, pkt.data[offset]) == 0x40
+end
+
+function is_ndp(pkt)
+   if pkt.length > 57 and
```
Comment (wingo) [inline src/apps/lwaftr/ndp.lua]: here too

File: src/apps/lwaftr/ndp.lua [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+       end
+   end
+   return false
+end
+
+function get_dst_ethernet(pkt, target_ipv6_addrs)
+   if pkt == nil or target_ipv6_addrs == nil then return false end
+   local na_addr_offset = eth_ipv6_size + 8
+   for i=1,#target_ipv6_addrs do
+      if C.memcmp(target_ipv6_addrs[i], pkt.data + na_addr_offset, 16) == 0 then
```
Comment (wingo) [inline src/apps/lwaftr/ndp.lua]: I think you want ipv6_equals() from lwaftr.lua.  I have to say also that we should be using more pointers into packets like the lwaftr does -- e.g. a pointer to the start of the ND payload start.  However I don't think that should block this merge.

File: src/program/lwaftr/setup.lua [hunk 5]
```diff
@@ -24,9 +26,13 @@ function lwaftr_app(c, conf)
...
    config.app(c, "fragmenterv6", ipv6_apps.Fragmenter,
               { mtu=conf.ipv6_mtu })
+   config.app(c, "ndp", ipv6_apps.NDP,
+              { src_ipv6 = conf.aftr_ipv6_ip, src_eth = conf.aftr_mac_b4_side,
+                dst_eth = conf.b4_mac, dst_ipv6 = conf.next_hop_ipv6_addr,
+                all_ipv6_addrs = conf.preloaded_binding_table:get_br_addresses() })
 
    local preprocessing_apps_v4  = { "reassemblerv4" }
-   local preprocessing_apps_v6  = { "reassemblerv6" }
+   local preprocessing_apps_v6  = { "reassemblerv6"}
```
Comment (wingo) [inline src/program/lwaftr/setup.lua]: can you keep the space here plz :)

**Comment by wingo** [comment]
The code is lovely, just lovely.  Before merging, would you mind commenting in the PR about how you've tested this?  I wouldn't think it would have a significant performance impact but you never know.  Also correctness of course. WDYT?

**Comment by kbara** [comment]
It's tested primarily by 6 new end-to-end tests (plus another 6 that are their vlan equivalents). Secondarily, it's tested with a selftest. It still needs testing against a real implementation. I believe all of the ICMP payloads are correct, but actual interoperability testing is always useful - having the same person write the code and tests (which I did) increases the odds of overlooking something.
Performance overheard should exist but not be huge: there's one extra check on incoming packets to see if they are NDP, and there's one new app. I strongly suspect we'll want to merge NDP/ping/etc into one app to reduce the overhead of a large chain of apps, but we need to set up appropriate benchmarking - it makes the code too much uglier to do it purely speculatively.
I'd recommend *against* using the feature to find the next hop's MAC address on start, because it drops packets until that's done (and if for whatever reason that fails, the only way to recover and transmit packets is to restart snabb-lwaftr). Using the feature to listen for NS and reply with solicited NAs should be fine. The tests are reasonably good, but not exhaustive. I expect quality is comparable to the pre-vlan code we had just before the alpha release.

File: src/program/lwaftr/setup.lua (hunk 5 above)
Comment (kbara) [inline src/program/lwaftr/setup.lua]: Yep, typo. Adrian'll be changing this for the changes in his ICMP patch anyhow, I think, but reverted. :)

**Comment by wingo** [comment]
Please test performance to make sure it's not leaking packets or anything -- just set up a normal run with transient on our servers.  Once that's done LGTM to merge :)  Thanks!

**Comment by kbara** [comment]
Run under full load with packet blasters; no sign of leaking packets. Thoroughput was improved by an amount that was almost certainly under the noise floor.

**Pull request merged**